staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog"
+)
+
+type Barrier interface {
+	SetComplete()
+	Completed() bool
+}
+
+var _ Barrier = &neverCompleteBarrier{}
+
+type neverCompleteBarrier struct{}
+
+func (_ *neverCompleteBarrier) SetComplete()    {}
+func (_ *neverCompleteBarrier) Completed() bool { return false }
+
+var _ Barrier = &StorageVersionBarrier{}
+
+// StorageVersionBarrier indicates if the aggregator, kube-apiserver, and the
+// apiextensions-apiserver have completed reporting their storage versions.
+type StorageVersionBarrier struct {
```

Comment:
if we made this hold all the StorageVersions that need to be registered, then we could selectively hold only requests for resources that don't yet have the storage versions registered.  Also, this could contain a controller fired by post-starthook that waits for a runonce to give a best shot at registration before becoming ready (tries once).  And I think it would simplify the registration process by allowing the installAPI to fully contain the detail of setting this to reconcile on its behalf.